[HLSL] Implement intangible AST type

hekota opened the pull request:
HLSL has a set of _intangible types_, which are described in in the [draft HLSL Specification (**[Basic.types]**)](https://microsoft.github.io/hlsl-specs/specs/hlsl.pdf):
> There are special implementation-defined types such as handle types, which fall into a category of standard intangible types. Intangible types are types that have no defined object representation or value representation, as such the size is unknown at compile time.
>
> A class type T is an intangible class type if it contains an base classes or members of intangible class type, standard intangible type, or arrays of such types. Standard intangible types and intangible class types are collectively called intangible types([9](https://microsoft.github.io/hlsl-specs/specs/hlsl.html#Intangible)).

This PR implements one standard intangible type `__builtin_hlsl_resource_t` and sets up the infrastructure that will make it easier to add more in the future, such as samplers or raytracing payload handles. The HLSL intangible types are declared in `clang/include/clang/Basic/HLSLIntangibleTypes.def` and this file is included with related macro definition in most places that require edits when a new type is added.

The new types are added as keywords and not typedefs to make sure they cannot be redeclared, and they can only be declared in builtin implicit headers. For example the `__builtin_hlsl_resource_t` type represents a handle to a memory resource and it is going to be used in builtin HLSL buffer types like this:

```
template <typename T>
class RWBuffer {
  [[hlsl::contained_type(T)]]
  [[hlsl::is_rov(false)]]
  [[hlsl::resource_class(uav)]]  
  __hlsl_resource_t Handle;
};
```

Part 1/3 of llvm/llvm-project#90631.

bogner committed:
wip: Stub out adding an HLSLResource builtin type

There are a couple of things that may be wrong here:

- Adding the PREDEF_TYPE to ASTBitCodes seems sketchy, but matches
  prior art.
- I skipped name mangling for now - can it come up?
- We use an unspellable name in a few places
- The type info matches `void *`. Does that make sense?

bogner committed:
wip: Name __builtin_hlsl_resource_t and give it a size for now

hekota committed:
Add HLSLIntangibleTypes.def and first HLSL intangible type HLSLResource.

Modify all places where HLSLResource needs to be added to use the def file.

Future intangible types will be added just to HLSLIntangibleTypes.def and everything else should just work.

Add getHLSLType virtual method to TargetCodeGenInfo to be overridden in future DirectXTargetCodeGenInfo class (#95952).

hekota committed:
Add typedef and simple AST test

hekota committed:
Make it a keyword instead of typedef so it cannot be redefined

Make it sizeless

Add tests

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into intangible-ast-type

hekota committed:
clang-format

hekota committed:
Merge branch 'main' into intangible-ast-type

hekota commented:
@ChuanqiXu9 - any idea why is your new test `Modules/no-external-type-id.cppm` failing on my PR?
It is expecting `// CHECK: <DECL_FUNCTION {{.*}} op8=4048` in the llvm-bcanalyzer output and it gets `op8=4056`.

python3kgae commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -2241,6 +2247,11 @@ TypeInfo ASTContext::getTypeInfoImpl(const Type *T) const {
 #include "clang/Basic/AMDGPUTypes.def"
+#define HLSL_INTANGIBLE_TYPE(Name, Id, SingletonId) case BuiltinType::Id:
+#include "clang/Basic/HLSLIntangibleTypes.def"
+      Width = 0;
```
Why the Width is 0?

hekota replied:
The intangible type is currently implemented as size-less type, but that might change as the HLSL implementation evolves. Setting the width to 0 here matches what similar types from other languages have done, for example `__externref_t` in WebAssembly, which is also an opaque type whose value cannot be accessed or manipulated, only passed around.

python3kgae commented on clang/lib/Index/USRGeneration.cpp:
```diff
@@ -757,7 +757,8 @@ void USRGenerator::VisitType(QualType T) {
         case BuiltinType::OCLSampler:
-          Out << "@BT@OCLSampler"; break;
+          Out << "@BT@OCLSampler";
+          break;
```
Is this expected?

hekota replied:
No, this is probably the clang-format side effect. I'll fix it.

hekota committed:
Cleanup

damyanp commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -115,6 +116,18 @@ GlobalVariable *replaceBuffer(CGHLSLRuntime::Buffer &Buf) {
+  if (llvm::Type *TargetTy = CGM.getTargetCodeGenInfo().getHLSLType(CGM, T))
+    return TargetTy;
+
+  // TODO: What do we actually want to do generically here? OpenCL uses a
+  // pointer in a particular address space.
+  llvm_unreachable("Generic handling of HLSL types is not implemented yet");
```
At what point will this become something we need to do?  Is there an issue tracking it?

hekota replied:
I am not sure if we have any HLSL specific type that will not be handled by `getHLSLType`, so this might just become simple `llvm_unreachable` without the TODO: comment. @bogner?

bogner replied:
For DirectX and SPIR-V we'll implement `getHLSLType` and this should really be unreachable, but we may want to do something generic in the future if/when we want to generate code for GPU or CPU targets directly generically. Something like the OpenCL approach mentioned in the comment might make sense in that case.

hekota replied:
I have removed the comment and left the `llvm_unreachable`. We can revisit this in the future.

damyanp commented:
I see many places where extra cases have been added for the intangible types but no corresponding tests.  Is that ok?  How did you know to update these places?

I also don't see anywhere that actually successfully uses `__builtin_hlsl_resource_t`. Am I missing it, or should I not expect to see it?

hekota commented:
> I see many places where extra cases have been added for the intangible types but no corresponding tests. Is that ok? How did you know to update these places?

I looked at similar types in other languages, such as the `image*` types in [OpenCL](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/OpenCLImageTypes.def) or `__externref_t` in [WebAssembly](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/WebAssemblyReferenceTypes.def). I used the same implementation style and defaults as these types (except `__externref_t` is a typedef and not a language keyword). Some of the cases are tested when the two `builtin_hlsl_resource_t.hlsl` files are parsed and AST dumped, but it is not possible to test all case until we can actually use the type (llvm/llvm-project#84824).

> I also don't see anywhere that actually successfully uses `__builtin_hlsl_resource_t`. Am I missing it, or should I not expect to see it?

You are correct - the type cannot be directly declared in user code, so the tests only check for errors and that the type shows up properly in AST. It is possible to use this type internally in Clang or in implicit headers though, which are two ways how we could define `RWBuffer` and other builtin types that will internally use `__builtin_hlsl_resource_t` handle. That will come alive in (llvm/llvm-project#84824).

llvm-beanz commented on clang/include/clang/Basic/HLSLIntangibleTypes.def:
```diff
@@ -0,0 +1,33 @@
+//
+//===----------------------------------------------------------------------===//
+
+HLSL_INTANGIBLE_TYPE(__builtin_hlsl_resource_t, HLSLResource, HLSLResourceTy)
```
I don't know if we have any cases where a type is prefixed by `__builtin`, usually types are `__clang` or  `__<target>` (e.g. `__arm`, `__amdgpu`). Maybe we should just do `__hlsl`?

llvm-beanz commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12335,6 +12335,10 @@ def warn_hlsl_availability_unavailable :
+def err_hlsl_intangible_type_cannot_be_declared : Error<
+"HLSL intangible type cannot be declared here">;
+def err_hlsl_intangible_type_as_function_arg_or_return : Error<
+"HLSL intangible type cannot be used as function %select{argument|return value}0">;
```
I don't think this is correct. We do allow resources as argument and return types.

llvm-beanz commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -3428,6 +3428,12 @@ void CXXNameMangler::mangleType(const BuiltinType *T) {
+#define HLSL_INTANGIBLE_TYPE(Name, Id, SingletonId)                            \
+  case BuiltinType::Id:                                                        \
+    type_name = #Name;                                                         \
+    Out << type_name.size() << type_name;                                      \
```
```suggestion
    Out << 'u' << type_name.size() << type_name;                               \
```

The `u` is significant here because it denotes a vendor-specific builtin type (see: https://itanium-cxx-abi.github.io/cxx-abi/abi-mangling.html).

llvm-beanz commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8001,6 +8001,12 @@ NamedDecl *Sema::ActOnVariableDeclarator(
 
+  if (getLangOpts().HLSL) {
+    if (R->isHLSLSpecificType() && !NewVD->isImplicit()) {
+      Diag(D.getBeginLoc(), diag::err_hlsl_intangible_type_cannot_be_declared);
```
Is the intent here to basically say that you can't create these types unless the declaration is implicit? I know we don't expose handle types in DXC, but do we really need to force these to be invalid in source?

hekota replied:
Yes, the intention was that these types can be created only in the implicit header `hlsl.h`. If that's not the case, I will remove this.

llvm-beanz replied:
`NewVD->isImplicit()` means the declaration is implicit and doesn't have any source, that prevents us from being able to use these in `hlsl.h` or directly in tests. I think that restriction may be more than we need.

hekota replied:
Got it, I thought `isImplicit()` also means declarations from implicit headers.

If we want to use these in `hlsl.h` then I don't think they can be truly sizeless. Sizeless types does not seem to be allowed in struct fields.

llvm-beanz replied:
HLSL's sizzles types will need to be allowed in struct fields, otherwise the property can't propagate correctly up to the resource objects or structures that contain resources which HLSL allows.

llvm-beanz reviewed (commented):
Few comments, but mostly I really like this direction!

bob80905 commented:
Just an FYI, the syntax for adding a resource_class attribute as currently implemented looks like this:
`[[hlsl::resource_class(SRV)]]`
In your description, you may want to update this and other similar attributes.

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into intangible-ast-type

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into intangible-ast-type

hekota committed:
Code review feedback

- rename to __hlsl_resource_t
- allow instantiation of sizeless intangible types in fields

Update expected type ID in no-external-type-id.cppm - the test does not have a stable baseline when new builtin types are added.

python3kgae commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -1390,7 +1390,8 @@ void ASTContext::InitBuiltinTypes(const TargetInfo &Target,
   }
 
-  if (Target.hasAArch64SVETypes()) {
+  if (Target.hasAArch64SVETypes() ||
```
Is this expected change for intangible AST type?

hekota replied:
I don't see this change in this PR diff. This change was done in llvm/llvm-project#99446 and came in via merge from `main`.

hekota committed:
clang-format - remove whitespace

hekota commented on clang/test/Modules/no-external-type-id.cppm:
```diff
@@ -23,7 +23,7 @@ export module b;
 import a;
 export int b();
 
-// CHECK: <DECL_FUNCTION {{.*}} op8=4048
```
@ChuanqiXu9 - FYI, I am updating the test baseline. The `op8` parameter is type ID and it is based on the number of builtin types. When a new type is added the test fails because the expected type ID value does not match anymore. Please consider updating the test baseline to be resistant against this. :)

hekota commented:
> > I also don't see anywhere that actually successfully uses `__builtin_hlsl_resource_t`. Am I missing it, or should I not expect to see it?
> 
> You are correct - the type cannot be directly declared in user code, so the tests only check for errors and that the type shows up properly in AST. It is possible to use this type internally in Clang or in implicit headers though, which are two ways how we could define `RWBuffer` and other builtin types that will internally use `__builtin_hlsl_resource_t` handle. That will come alive in (#84824).

The PR has been updated to allow instantiation of `__hlsl_resource_t` in struct fields.

hekota committed:
[NFC] revert formatting

hekota committed:
Merge branch 'main' into intangible-ast-type

hekota committed:
Merge branch 'main' into intangible-ast-type

bogner reviewed (approved):
LGTM!
